core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -161,49 +178,68 @@ private[chisel3] trait HasId extends InstanceId {
     // This could be called with user prefixes, ignore them
     noPrefix {
       suggested_seed = Some(seed)
-      this.suggestName(seed)
+      this.suggestNameInternal(seed)
     }
   }
 
+  private[chisel3] var _computedName: Option[Option[String]] = None
+
   /** Computes the name of this HasId, if one exists
     * @param defaultPrefix Optionally provide a default prefix for computing the name
     * @param defaultSeed Optionally provide default seed for computing the name
     * @return the name, if it can be computed
     */
   private[chisel3] def _computeName(defaultPrefix: Option[String], defaultSeed: Option[String]): Option[String] = {
 
-    /** Computes a name of this signal, given the seed and prefix
-      * @param seed
-      * @param prefix
-      * @return
-      */
-    def buildName(seed: String, prefix: Prefix): String = {
-      val builder = new StringBuilder()
-      prefix.foreach { p =>
-        builder ++= p
-        builder += '_'
+    _computedName.getOrElse {
+
+      /** Computes a name of this signal, given the seed and prefix
+        * @param seed
+        * @param prefix
+        * @return
+        */
+      def buildName(seed: String, prefix: Prefix): String = {
+        val builder = new StringBuilder()
+        prefix.foreach { p =>
+          builder ++= p
+          builder += '_'
+        }
+        builder ++= seed
+        builder.toString
       }
-      builder ++= seed
-      builder.toString
-    }
 
-    if (hasSeed) {
-      Some(buildName(seedOpt.get, naming_prefix.reverse))
-    } else {
-      defaultSeed.map { default =>
-        defaultPrefix match {
-          case Some(p) => buildName(default, p :: naming_prefix.reverse)
-          case None    => buildName(default, naming_prefix.reverse)
+      val result = if (hasSeed) {
+        Some(buildName(seedOpt.get, naming_prefix.reverse))
+      } else {
+        defaultSeed.map { default =>
+          defaultPrefix match {
+            case Some(p) => buildName(default, p :: naming_prefix.reverse)
+            case None    => buildName(default, naming_prefix.reverse)
+          }
         }
       }
+      _computedName = Some(result)
+      result
     }
   }
 
   /** This resolves the precedence of [[autoSeed]] and [[suggestName]]
+    *
+    * @note It will become an error in the future to suggestName the same thing that autoSeed would have assigned
     *
     * @return the current calculation of a name, if it exists
     */
-  private[chisel3] def seedOpt: Option[String] = suggested_seed.orElse(auto_seed)
+  private[chisel3] def seedOpt: Option[String] = {
+    suggested_seed.zip(auto_seed).foreach {
+      case (suggested, auto) =>
+        if (suggested == auto) {
+          Builder.deprecated(
+            s"calling suggestName(${suggested}) had no effect as it is the same as the auto prefixed name, this will become an error in 3.6"
+          )
+        }
+    }
+    suggested_seed.orElse(auto_seed)
+  }
```

Comment:
I don't think this is the right place to put this since, internally it seems reasonable to call `seedOpt` multiple times but now it might add a warning multiple times. Can we instead issue the warning upon calling `.suggestName`?

Reply:
no... because at time of suggestName we don't know what the auto seed is?